tsfc/loopy.py

```diff
@@ -40,26 +43,67 @@ def pym_multiindex(self, multiindex):
                 indices.append(index)
         return tuple(indices)
 
-    def pymbolic_variable(self, node):
+    def gem_to_pym(self, node):
         try:
             pym = self.gem_to_pymbolic[node]
         except KeyError:
             name = self.name_gen(node.name)
             pym = p.Variable(name)
             self.gem_to_pymbolic[node] = pym
+        return pym
+
+    def pymbolic_variable(self, node):
+        pym = self.gem_to_pym(node)
         if node in self.indices:
             indices = self.pym_multiindex(self.indices[node])
             if indices:
                 return p.Subscript(pym, indices)
-            else:
-                return pym
-        else:
-            return pym
+        return pym
+
+    def pymbolic_variable_and_destruct(self, node):
+        pym = self.gem_to_pym(node)
+        if node in self.indices:
+            indices = self.pym_multiindex(self.indices[node])
+            if indices:
+                return (pym, indices)
+        return (pym, ())
 
     def active_inames(self):
         # Return all active indices
         return frozenset([i.name for i in self.active_indices.values()])
 
+    def pymbolic_indices(self, expr):
+        try:
+            indices = expr.multiindex
+        except AttributeError:
+            indices = tuple(shp for shp in expr.shape)
+        pym_idx = tuple()
+        for index in indices:
+            if type(index) == gem.Index:
+                if not index.name:
+                    name = self.name_gen(self.index_names[index])
+                else:
+                    name = index.name
+                assert index.extent
+                extent = index.extent
+            else:
+                name = self.name_gen(self.index_names[index])
+                extent = index
+            self.index_extent[name] = extent
+            pym_idx += (p.Variable(name), )
+        return pym_idx
+
+
+@contextmanager
```

Comment:
This is only available from python3.7 and we test Tsfc with python3.5. Should I use something else for the context manager?

Reply:
`contextmanager` has been around at least since Python 2.7

Reply:
Oh what, okay, seems like I misread google. Not quite sure why the branch does not build then.

Reply:
```
pip._vendor.urllib3.exceptions.ProtocolError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
288The command "pip install -r requirements.txt" failed and exited with 2 during .``` You had a network failure on travis.